rancher/dashboard
# Add catalog list/loading for air-gapped extensions

### Summary

This adds a new option to import Extension Catalogs from a specific registry image and load multiple packages as unique charts.

#### Additions
- "Extension Catalog" list view
  - Contains ResourceList for loaded extension catalog images
- Dialog for loading image with secrets
- Github workflow for building  Extension container image

#### Loading process
The catalog loading process consists of 4 parts:
- Creating the deployment from the extension image
- Exposing the deployment with a ClusterIP service
- Creating a Helm repository from the service's ClusterIP
- Creating the UIPlugin crd for the Extension.

Once all of the resources have been successfully created, the available charts provided by the new Helm repository will be shown in the "Available" tab.

#### Extension build process
The build process for the container image of an extension has been modified to allow the root Extension to act as a Helm repository and each package within the `./pkg` directory to be listed as charts. 

Also the dependency on the [ui-plugin-server](https://github.com/rancher/ui-plugin-server) has been removed and placed into the `./shell/scripts/extension/helm` directory. Instead of cloning the repository during the `publish` script, the contents of the `helm` directory is copied and iterated over for each package. This directory contains a condensed version of the chart originally used (`./helm/charts`), the Dockerfile and nginx.conf for the image build (`./helm/package`), and the scripts for used to populate the config for the build (`./helm/scripts`).

#### Extension workflow
A new workflow (`build-container.yml`) has been added to the `@rancher/pkg` creator. This will build the extension as a container image and push it into the Github Container Registry (`ghcr.io`). This will make the image easily accessible for a user to tag and push into a specified registry.  

### To test
Follow the documentation in `./docusaurus/docs/extensions/advanced/air-gapped-environments.md`. This [container image](https://github.com/jordojordo/extensions-test-pt-deux/pkgs/container/ui-extension-extensions-test-pt-deux/94081344?tag=0.1.2) can be used to test, it contains 2 top-level extension packages. If necessary, I can provide private registry as well.

1. Navigate to Catalog list by selecting the "Manage Extension Catalog" action item
![manage-catalog](https://github.com/rancher/dashboard/assets/40806497/63c26194-df5c-41ec-b3b0-029127446d14)

2. Import Extension Catalog with the new dialog 
![load](https://github.com/rancher/dashboard/assets/40806497/c317a22b-e5dd-44ca-899c-2dec819799ff)

3. See cache state updated for image
![cached](https://github.com/rancher/dashboard/assets/40806497/0cee0541-f80c-4469-9be4-ac7d57590c60)

4. Navigate back to the chart list from the "Manage Extension Chart" action item
![manage-chart](https://github.com/rancher/dashboard/assets/40806497/a819e503-b107-4b88-8bab-f8ab17c95b10)

5. Install one of the new extensions in the Available tab
6. Once installed navigate to the new menu item within the side-nav
